distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso

```diff
@@ -0,0 +1,166 @@
+
+## PRIVATE
+   Returns an access token for the current user, that should be valid for
+   at least 2 more minutes.
+
+   If the current token is expired or <2m from expiration, it will be refreshed
+   and a new one will be returned. Because of that, this method may make network
+   requests.
+get_access_token : Text
+get_access_token = AuthenticationProvider.getAccessToken
+
+## PRIVATE
+   Forcibly refreshes the access token.
+refresh_access_token : Nothing
+refresh_access_token =
+    AuthenticationProvider.getAuthenticationServiceEnsoInstance.force_refresh
+
+## PRIVATE
+credentials_file : File
+credentials_file = case Environment.get "ENSO_CLOUD_CREDENTIALS_FILE" of
+    Nothing -> File.home / ".enso" / "credentials"
+    path -> File.new path
+
+## PRIVATE
+type Authentication_Service
+    ## PRIVATE
+    Instance (auth_data : Ref Authentication_Data)
+
+    ## PRIVATE
+    get_access_token self -> Text =
+        is_still_valid = self.auth_data.get.expire_at > (Date_Time.now + token_early_refresh_period)
+        if is_still_valid then self.auth_data.get.access_token else
+            # The token has expired or will expire soon, so we need to refresh it.
+            self.force_refresh
+            self.auth_data.get.access_token
+
+    ## PRIVATE
+       Refreshes the access token using the refresh token, regardless of its freshness.
+    force_refresh self -> Nothing =
+        refresh_token_data = Refresh_Token_Data.read_from_credentials
+        new_auth_data = refresh_token_data.fetch_new_access_token
+        Panic.rethrow <| self.auth_data.put new_auth_data
+        Nothing
+
+    ## PRIVATE
+    new -> Authentication_Service =
+        initial_data = Authentication_Data.read_from_credentials
+        Authentication_Service.Instance (Ref.new initial_data)
+
+## PRIVATE
+   The method called from Java that creates a new instance of the service, to be
+   stored in the global state.
+instantiate_authentication_service =
+    # We rethrow any dataflow errors, because this is fed back to Java code that only understands exceptions (panics).
+    Panic.rethrow <|
+        Authentication_Service.new
+
+## PRIVATE
+type Authentication_Data
+    ## PRIVATE
+    Value access_token:Text expire_at:Date_Time
+
+    ## PRIVATE
+    read_from_credentials -> Authentication_Data =
+        content = credentials_file.read_text . catch File_Error error-> case error of
+            File_Error.Not_Found _ -> Panic.throw Not_Logged_In.Error
+            _ -> Panic.throw (Illegal_State.Error "Failed to read credentials file: "+error.to_display_text cause=error)
+        as_json = content.parse_json.catch Invalid_JSON error->
+            Panic.throw (Illegal_State.Error "Invalid credentials file format: "+error.to_display_text cause=error)
+        get_field field_name =
+            as_json.get field_name if_missing=(Panic.throw (Illegal_State.Error "Invalid credentials file format: missing field `"+field_name+"`."))
```

Comment:
This could be shared with the similar code in `Refresh_Token_Data`, passing in a prefix for error messages.

Reply:
IMO it's such a small piece of code that abstracting it and adding the prefix makes the code more complicated.

Reply:
After second thought I will try to improve the error handling a bit.

Reply:
Refactored in https://github.com/enso-org/enso/pull/9390/commits/d87232886ab733cc267d0f0ed55bc1cd980f2eb2